ftprime/argrecorder.py

```diff
@@ -1,48 +1,117 @@
+        """
+        :param ts TreeSequence: The tree sequence defining relationships between
+            the first genreation of individuals.
+        :param list first_generation: The input IDs of the first generation that
+            should be mapped onto the samples in the tree sequence ``ts``.
+        :param float time: The (forwards) time of the "present" at the start of
+            the simulation.
+        :param int max_indivs: The largest number of individuals (past or present)
+            that we will track at any given time.
+        """
+        # maximum number of individuals that can be kept track of in the internal state
+        self.n_max = max_indivs
+        # this is output node ID of the next new node
+        self.num_nodes = ts.num_nodes # n
+        # this is the largest (forwards) time seen so far
+        self.max_time = time  # T
+        # last (forwards) time we updated node times
+        self.last_update_time = time  # T_0
+        # number of nodes that have the time right
+        self.last_update_node = ts.num_nodes
+        assert len(ts.samples()) == len(first_generation)
+        # minimum currently active input ID
+        self.input_min = 0  # m_0
+        # list of output node IDs indexed by input labels
+        self.node_ids = np.empty(self.n_max, dtype='int32') # L
+        self.reset_node_ids(first_generation, ts.samples())
+        # the actual tables that get updated
+        #  DON'T actually store ts, just the tables:
+        self.nodes = msprime.NodeTable()
+        self.edgesets = msprime.EdgesetTable()
+        self.sites = msprime.SiteTable()
+        self.mutations = msprime.MutationTable()
+        self.tables = ts.dump_tables(nodes=self.nodes, edgesets=self.edgesets,
+                                     sites=self.sites, mutations=self.mutations)
 
     def __str__(self):
-        ret = self.node_table().__str__()
+        ret = "Input ID offset m_0:\n"
+        ret += str(self.input_min)
+        ret += "Max time so far:\n"
+        ret += str(self.max_time)
+        ret += "Last update time:"
+        ret += str(self.last_update_time)
+        ret += "Node IDs:\n"
+        ret += self.node_ids.__str__()
+        ret += "Tables:\n"
+        for x in self.tables:
+            ret += x.__str__()
         ret += "\n---------\n"
-        ret += self.edgeset_table().__str__()
         return ret
 
-    def add_individual(self, name, time, population=msprime.NULL_POPULATION,
-                       is_sample=False):
+    def reset_node_ids(self, generation, samples):
+        """
+        Reset the internal mapping of input to node IDs except for each
+        individual in generation is mapped to the corresponding one in samples.
+        """
+        self.node_ids.fill(NULL_ID)
+        for j, x in zip(generation, samples):
+            self.node_ids[self.input_min + j] = x
+
+    def add_individual(self, name, time, population=msprime.NULL_POPULATION):
```

Comment:
I would usually expect ``name`` to be a string, so had to read over this a few times to make sure I understood what's happening. Perhaps ``input_id`` or something would be a closer fit?